flang/lib/Lower/OpenMP.cpp

```diff
@@ -426,14 +425,10 @@ void DataSharingProcessor::collectSymbols(
                              /*collectSymbols=*/true,
                              /*collectHostAssociatedSymbols=*/true);
   for (Fortran::lower::pft::Evaluation &e : eval.getNestedEvaluations()) {
-    if (e.hasNestedEvaluations())
+    if (e.hasNestedEvaluations() && !e.isConstruct())
```

Comment:
Why is it collected in non-constructs (like OpenMP)?

Reply:
@kiranchandramohan I was on leave mainly for past few days; sorry about the delayed response.

So we maintain two lists: `defaultSymbols` and `symbolsInNestedRegions`. For non-constructs like OpenMP, we would not like to collected the host associated symbols for nested regions (thereby `collectHostAssociatedSymbols = false`). They can be handled as the `default` clause may be in their respective nested blocks. This was the earlier design.

However, for non-OpenMP constructs, the handling is different. This is because even though the inner construct has a block region (like OpenMP constructs do), all privatization of variables in the nested construct needs to happen in the outer OpenMP construct. Hence, we do not need to populate the `symbolsInNestedRegions` list. Otherwise, no privatization will happen, leading to incorrect results.

That said, if you think so, I can relook the design of the lists once. It has been quite some time when we implemented the default clause. So maybe, we can think of improving this lowering flow once. What are your suggestions?

Reply:
@kiranchandramohan A ping on this! How can we take this forward?

Reply:
> So we maintain two lists: defaultSymbols and symbolsInNestedRegions. For non-constructs like OpenMP, we would not like to collected the host associated symbols for nested regions (thereby collectHostAssociatedSymbols = false). They can be handled as the default clause may be in their respective nested blocks. This was the earlier design.

Some OpenMP constructs (like critical) will not have a default clause because such a clause is not supported for these constructs. Some OpenMP constructs might not have them specified (i.e no default clause is present eventhough it is supported). Do we need to distinguish these cases? For those constructs which have them specified, for every pair of defaults do their exact values matter for the inner construct's symbols?

outer(default=shared), inner(default=shared) : Nothing to do
outer(default=private), inner(default=shared) : ?
outer(default=shared), inner(default=private) : ?
outer(default=private),inner(default=private) : ?

> However, for non-OpenMP constructs, the handling is different. This is because even though the inner construct has a block region (like OpenMP constructs do), all privatization of variables in the nested construct needs to happen in the outer OpenMP construct. Hence, we do not need to populate the symbolsInNestedRegions list.

Makes sense.

Reply:
@kiranchandramohan I checked for the use-cases mentioned. The lowering seems to be okay. Should I add these combinations in the tests for default clause?

**outer(default=private), inner(default=private)**

Consider the following example:

```
integer :: j
!$omp parallel default(private)
      j = 10
      !$omp parallel default(firstprivate)
                 j = j + 5
      !$omp end parallel
!$omp end parallel
```

This lowers as:

```
%0 = fir.alloca i32 ......
%1:2 = hlfir.declare %0 .....
omp.parallel {
       %2 = fir.alloca i32 .....      // privatized j
        %3:2 = hlfir.declare %2
        %c = arith.constant 10 
        hlfir.assign %c to %3#0 ....
        omp.parallel {
                 %4 = fir.alloca i32 ....
                 %5:2 = hlfir.declare %4 ....
                 %6 = fir.load %3#0 .....
                 hlfir.assign %6 to %5#0 ......
        }
}
```

**outer(default=private), inner(default=shared)**

Example:

```
integer :: j
!$omp parallel default(private)
      j = 10
      !$omp parallel
                 j = j + 5
      !$omp end parallel
!$omp end parallel
```

This lowers as:

```
%0 = fir.alloca i32 ......
%1:2 = hlfir.declare %0 .....
omp.parallel {
       %2 = fir.alloca i32 .....      // privatized j
        %3:2 = hlfir.declare %2
        %c = arith.constant 10 
        hlfir.assign %c to %3#0 ....
        omp.parallel {
                 %4 = fir.load %3#0 .....
                 // add 5 to %4 and store in %5
                 hlfir.assign %5 to %3#0 ....
        }
}
```

**outer(default=shared), inner(default=private)**

Example:

```
integer :: j
!$omp parallel
      j = 10
      !$omp parallel default(private)
                 j = j + 5
      !$omp end parallel
!$omp end parallel
```

This lowers as:

```
%0 = fir.alloca i32 ......
%1:2 = hlfir.declare %0 .....
omp.parallel {
        %c = arith.constant 10 
        hlfir.assign %c to %3#0 ....
        omp.parallel {
                 %2 = fir.alloca ....
                 %3:2 = hlfir.declare %2 ....
                 %4 = fir.load %3#0 .....
                 // add 5 to %4 and store in %5
                 hlfir.assign %5 to %3#0 ....
        }
}
```

And should be consider `firstprivate` in the inner parallel:

```
%0 = fir.alloca i32 ......
%1:2 = hlfir.declare %0 .....
omp.parallel {
        %c = arith.constant 10 
        hlfir.assign %c to %3#0 ....
        omp.parallel {
                 %2 = fir.alloca ....
                 %3:2 = hlfir.declare %2 ....
                 %4 = fir.load %1#0 ....
                 hlfir.assign %4 to %3#0 .....
                 %5 = fir.load %3#0 .....
                 // add 5 to %4 and store in %6
                 hlfir.assign %6 to %3#0 ....
        }
}
```

Reply:
Additionally, for constructs like `critical` not having an explicit default clause, the lowered IR is in line with the correct symbols that the construct should be using.

Reply:
I was trying to understand the need to distinguish between OpenMP and non-OpenMP constructs for collecting symbols recursively.  Particularly in cases where the variable on which a default applies is only visible inside the nested OpenMP construct. Consider the following example : 

```
program mn
  call sb1()
contains
 subroutine sb1()
   integer :: x
   x = 20
   !$omp parallel default(private)
   !$omp critical
   x = 30
   !$omp end critical
   !$omp end parallel
   print *, x
 end subroutine
end program
```

But the following example with nested parallel works. I am assuming the semantics code is creating symbols here. Is that the reason? 
If so, why not handle everything (nested openmp and nested non-openmp constructs) in the frontend? What is the split between semantics and the code here?
```
program mn
  call sb1()
contains
 subroutine sb1()
   integer :: x
   x = 20
   !$omp parallel default(private)
   !$omp parallel 
   x = 30
   !$omp end parallel
   !$omp end parallel
   print *, x
 end subroutine
end program
```

Reply:
Hello @kiranchandramohan,

        I  re-looked at the possibility of handling the entire default clause at one place. For instance, I checked whether we can handle it completely during lowering. One possible way:

1. Use a function similar to `collectSymbolSet` with a bool `checkFlag`, such that the function checks symbol flag iff `checkFlag = True`, otherwise it simply adds the symbol to the symbol list.

2. Given any OpenMP context, all symbols encountered in the same context can then be collected through `checkFlag = True`. In other words, any symbol in the same region where `default(private)` is specified should only collect symbols with the `private` data-sharing attribute. However, any symbol in a nested region can be collected irrespective of the specified flag on that symbol.

This approach, unfortunately, does not work without the symbol modifications we currently do in semantics.

For your question on handling OpenMP and non OpenMP constructs in semantics itself, currently, both are being handled. It's just that while lowering, for nested OpenMP constructs, we need to collect host-associated symbols as well. For non-OpenMP constructs, we skip host-associated symbols.

I tried a few variations in which to simply this (use a single list `defaultSymbols` instead of two lists: `defaultSymbols` and `symbolsInNestedRegions`). But so far, nothing concrete has come up...